### DISCUSSION TITLE
PDF form controls issues in latest versions

### DISCUSSION THREAD

**Discussion by GopiUC** [stephanrauh/ngx-extended-pdf-viewer#2121] [Q&A]
Hi,

1.CheckBox saved as unchecked after edit.
2. Selected radio button is not selected in edit mode.

**1**. In few Checkboxes checked and saved the PDF form successfully.

![CheckboxIssue2](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/158054442/c06cb247-9ce1-4003-8f1d-f750a8152fe0)

In our functionality, we can edit the PDF form, 
Check other checkboxes, and save it again. However, the form data sends value as 'undefined ' for previously checked boxes. 
Only newly selected check boxes are saved as 'true' in the pdf, while those already checked are unchecked.

*This functionality working fine in Version 16.2.8.

![CheckBoxIssue1](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/158054442/10a7822b-19e5-4b39-b835-aa84ee3e1b5e)

**2**. And another issue we faced is, From version 18.1.1 - In Edit mode Selected radio button is seems like unselected, but in view mode it displayed correctly what we fetched from API. Till 18.1.0 it works fine.  So, now we roll back to 18.1.0.

Both functionalities are not correctly working in current versions. Please check and advise how can we fix this issues.
Checkbox issue is very crucial for us.

Thank you for your time,
GopiVignesh

**Comment by stephanrauh** [comment]
I wonder if this is a duplicate of #2052?

**Reply by stephanrauh** [reply]
You're referring to an "edit mode". Can you explain what this is? The PDf viewer does not distinguish between an edit mode and an read-only mode - it always allows users to check and uncheck checkboxes.

**Reply by GopiUC** [reply]
For first point 
I thought issue not like that bug #2052.
Initially checkbox loaded properly but during update of the PDF form the formdata not updated properly. The formData sends value as 'undefined ' for previously checked boxes. Refer the second image of previous message.

**Reply by GopiUC** [reply]
For the second point, 'edit mode' and 'view mode' refer to the 'FormFlattening' flag 'false' and 'true' respectively. 'FormFlattening' is the property of pdfStamper. That FormFlattening flag handled in GetAPI of PDF form. 
That flag was used to determine whether the form should be used for editing or viewing only.

![RadioButtonIssue1](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/158054442/354b3d1c-df47-4197-91fb-8bc4159461df)

Please refer above img, we used radio button like this.

**Reply by GopiUC** [reply]
I wanted to kindly follow up on my previous request for clarification on those points. If you could spare a moment to review and provide your insights, I would greatly appreciate it. Your input is invaluable in moving the project forward efficiently.

Thank you for your attention to this matter. Looking forward to your response.

**Reply by stephanrauh** [reply]
I can't reproduce your bug on https://pdfviewer.net/extended-pdf-viewer/forms. Maybe it has something to do with your particular form. However, I wonder why you fetch the form data using `pdfService.getFormData()`? That's a low-level API giving you the raw data. Is there a reason why you don't use the data passed in `$event`?

**Reply by GopiUC** [reply]
Thank you for the response, I used $event data to update formdata. Now chcekBox value updated correctly.

**Reply by GopiUC** [reply]
For Second Point, In that form Radio button used like checkbox. Please check that point too and refer first message

"2. And another issue we faced is, From version 18.1.1 - In Edit mode Selected radio button is seems like unselected, but in view mode it displayed correctly what we fetched from API. Till 18.1.0 it works fine. So, now we roll back to 18.1.0."

"For the second point, 'edit mode' and 'view mode' refer to the 'FormFlattening' flag 'false' and 'true' respectively. 'FormFlattening' is the property of pdfStamper. That FormFlattening flag handled in GetAPI of PDF form.
That flag was used to determine whether the form should be used for editing or viewing only.

![RadioButtonIssue1](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/158054442/354b3d1c-df47-4197-91fb-8bc4159461df)

Please refer above img, we used radio button like this."

**Reply by stephanrauh** [reply]
Sorry for responding so sluggishly! I've tried to find out what I changed between version 18.1.0 and 18.1.1. I don't see many form-related changes. Maybe this commit causes trouble: https://github.com/stephanrauh/pdf.js/commit/a6163a2896c8308b10af28067bb26d9a7db31344. 

You've sent me only one screenshot, so I can only guess whether that's the correct or the faulty state. When you say "the radio button seems like unselected", is it only a display problem, or is it disabled in the DOM, too? You can check this using the "elements" tool of your browser's developer tools. I also wonder if you PDF form is a traditional AcroForm or an XFA form.

**Reply by GopiUC** [reply]
Thanks for your response and efforts in investigating the issue.

Confirming that the radio button appears disabled both visually and in the DOM when we get form for update. I've checked using browser's developer tools. In version 18.1.1

Regarding the PDF form type, it's a traditional AcroForm only.

![RadioButtonIssue2](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/158054442/cfdf8580-b46a-4b0c-9443-561d9aef52a6)

In this screenshot that radio button was selected and saved. When get form for edit It seems like unselected, in DOM too. 

But I open that saved form in WPS office it seems selected image attached below.

![RadioButtonIssue3](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/158054442/7bda44d6-97b2-4475-bb5f-adefd3c42697)
